File: sre-agent/internal/observability/loki_direct.go
package observability

import (
	"context"
	"encoding/base64"
	"fmt"
	"net/url"
	"strings"
	"time"
)

// LokiDirectAdapter queries Loki directly via its HTTP API.
type LokiDirectAdapter struct {
	URL       string
	OrgID     string
	Tenant    string
	BasicAuth *BasicAuth
}

// BasicAuth holds optional basic auth credentials.
type BasicAuth struct {
	User string
	Pass string
}

// NewLokiDirectAdapter creates a new Loki direct adapter.
func NewLokiDirectAdapter(rawURL, orgID, tenant, user, pass string) *LokiDirectAdapter {
	a := &LokiDirectAdapter{
		URL:    strings.TrimSuffix(rawURL, "/"),
		OrgID:  orgID,
		Tenant: tenant,
	}
	if user != "" || pass != "" {
		a.BasicAuth = &BasicAuth{User: user, Pass: pass}
	}
	return a
}

// QueryErrors queries Loki's query_range endpoint for error logs.
func (a *LokiDirectAdapter) QueryErrors(ctx context.Context, timeWindow string) ([]ErrorLog, error) {
	end := time.Now().UnixNano()
	startOffset, err := parseDuration(timeWindow)
	if err != nil {
		return nil, fmt.Errorf("parse time window %q: %w", timeWindow, err)
	}
	start := time.Now().Add(-startOffset).UnixNano()

	query := BuildLogQL(FilterConfig{
		ErrorPattern: `level=~"(?i)error|fatal|panic"`,
	})

	u := url.Values{}
	u.Set("query", query)
	u.Set("start", fmt.Sprintf("%d", start))
	u.Set("end", fmt.Sprintf("%d", end))
	u.Set("step", "60")

	fullURL := fmt.Sprintf("%s/loki/api/v1/query_range?%s", a.URL, u.Encode())

	data, err := a.get(ctx, fullURL)
	if err != nil {
		return nil, fmt.Errorf("loki query: %w", err)
	}

	return ParseLokiResponse(data)
}

// TestConnection checks if Loki is reachable.
func (a *LokiDirectAdapter) TestConnection(ctx context.Context) error {
	u := fmt.Sprintf("%s/ready", a.URL)
	_, err := a.get(ctx, u)
	return err
}

// get performs a GET request with auth headers.
func (a *LokiDirectAdapter) get(ctx context.Context, rawURL string) ([]byte, error) {
	headers := map[string]string{}
	if a.Tenant != "" {
		headers["X-Scope-OrgID"] = a.Tenant
	}
	if a.BasicAuth != nil {
		creds := base64.StdEncoding.EncodeToString(
			[]byte(a.BasicAuth.User + ":" + a.BasicAuth.Pass))
		headers["Authorization"] = "Basic " + creds
	}
	return doGET(ctx, rawURL, headers)
}

// parseDuration handles Loki time windows like "30m", "2h", "1d", "5m".
func parseDuration(s string) (time.Duration, error) {
	d, err := time.ParseDuration(s)
	if err == nil {
		return d, nil
	}
	// Try bare minutes
	var mins int
	if _, err := fmt.Sscanf(s, "%dm", &mins); err == nil {
		return time.Duration(mins) * time.Minute, nil
	}
	var hours int
	if _, err := fmt.Sscanf(s, "%dh", &hours); err == nil {
		return time.Duration(hours) * time.Hour, nil
	}
	return 0, fmt.Errorf("unknown duration format: %s", s)
}

File: framework/ops_phoenix/adapters/loki_direct.py
"""Loki direct query adapter."""

import json
import os
from typing import Any, Dict, List

from ops_phoenix.adapters.base import ObservabilityAdapter


class LokiDirectAdapter(ObservabilityAdapter):
    """Query Loki directly via HTTP API."""

    def __init__(self, config: dict):
        super().__init__(config)
        self.url = config.get("url", "http://localhost:3100")
        self.org_id = config.get("org_id", "1")
        self.tenant = config.get("tenant")

    def query_errors(self, time_window: str = "1h") -> List[str]:
        """Query Loki for error log lines."""
        import time as time_mod
        from datetime import datetime, timedelta
        import urllib.request, urllib.parse

        end_ns = time_mod.time_ns()
        start = datetime.now() - timedelta(hours=1)
        start_ns = int(start.timestamp() * 1e9)
        duration_ns = end_ns - start_ns

        expr = '{{severity=~"(?i)error|fatal|panic"}}'
        expr += self.get_container_filter()

        url = (
            f"{self.url}/loki/api/v1/query_range"
            f"?query={urllib.parse.quote(expr)}"
            f"&start={start_ns}&end={end_ns}&step=60"
        )

        req = urllib.request.Request(url)
        if self.tenant:
            req.add_header("X-Scope-OrgID", self.tenant)

        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                data = json.loads(resp.read())
                results = data.get("data", {}).get("result", [])
                return [str(r) for r in results]
        except Exception as e:
            print(f"Loki query failed: {e}")
            return []

    def test_connection(self) -> bool:
        """Check if Loki is reachable."""
        try:
            import urllib.request
            req = urllib.request.Request(f"{self.url}/ready")
            if self.tenant:
                req.add_header("X-Scope-OrgID", self.tenant)
            with urllib.request.urlopen(req, timeout=10) as resp:
                return resp.status == 200
        except Exception:
            return False
